pkg/lib/cockpit-upload-helper.js

```diff
@@ -0,0 +1,164 @@
+            binary: true,
+            payload: "fsreplace1",
+            path: this.destination,
+            "send-acks": "frames",
+            ...options,
+        });
+        this.channel.addEventListener("control", this.on_control.bind(this));
+    }
+
+    // Private methods
+    on_control(event, message) {
+        if (message?.command === "ack") {
+            this.outstanding -= message.frames;
+
+            console.assert(this.outstanding >= 0, "outstanding went negative", this.outstanding);
+
+            if (this.resolveHandler) {
+                this.resolveHandler();
+                this.resolveHandler = null;
+            }
+        }
+    }
+
+    read_chunk(blob) {
+        return new Promise((resolve, reject) => {
+            const reader = new FileReader();
+            reader.onload = (event) => resolve(event.target.result);
+            reader.onerror = reject;
+            reader.readAsArrayBuffer(blob);
+        });
+    }
+
+    write(chunk) {
+        console.assert(this.outstanding >= 0, "ack outstanding went negative", this.outstanding);
+        this.outstanding += 1;
+        this.channel.send(chunk);
+
+        if (this.outstanding < TOTAL_FRAME_WINDOW) {
+            return Promise.resolve();
+        } else {
+            return new Promise((resolve, reject) => {
+                this.resolveHandler = resolve;
+            });
+        }
+    }
+
+    close() {
+        const channel = this.channel;
+        return new Promise((resolve, reject) => {
+            channel.addEventListener("close", function(event, message) {
+                if (message.problem) {
+                    reject(message);
+                } else {
+                    resolve();
+                }
+            });
+            channel.control({ command: "done" });
+        });
+    }
+
+    wait_ready() {
+        const channel = this.channel;
+        return new Promise((resolve, reject) => {
+            channel.addEventListener("ready", function(event, message) {
+                resolve();
+            });
+        });
+    }
+
+    // Public methods
+
+    // Flush is waiting on acks in flight
+    async upload(file) {
+        let close_message = null;
+        this.channel.addEventListener("close", (event, message) => {
+            // If we are waiting on an "ack" and the disk is full
+            if (this.resolveHandler)
+                this.resolveHandler();
+            close_message = message;
+        });
+
+        await this.wait_ready();
+
+        let chunk_start = 0;
+        let send_chunks = 0;
+
+        while (this.channel.valid && chunk_start <= file.size) {
+            const chunk_next = chunk_start + BLOCK_SIZE;
+            const blob = file.slice(chunk_start, chunk_next);
+
+            const chunk = await this.read_chunk(blob);
+            await this.write(chunk);
+
+            send_chunks += blob.size;
+            if (this.progressCallback)
+                this.progressCallback(send_chunks);
```

Comment:
In the hypothetical case we were discussing before about uploading 2MB over a slow connection, this would result in the progress bar immediately jumping to 100% and then stalling there.

Would it be possible to move this to after ack gets received?

Reply:
The acks have no idea about how much bytes were send, so we would have to let write somehow track that? I don't think this is feasible until you return bytes as `ack` message.